File: examples/RUNES_demo/robot/AODV/hellocode.cpp
double hellocode(int seg, void* data) {



  Hello_Data *d = (Hello_Data*) data;



  int activeroute = 0;

  int k, m, n;

  double now;

  double period;



  HelloMsg *hellomsg;

  GenericNwkMsg *nwkmsg;

  RERRMsg *rerrmsg;

  DataNode *dn;

  DataNode *tmp;

  

  switch (seg) {

  case 1:

    now = ttCurrentTime();

  

    if (VERBOSE) {

      mexPrintf("Time: %f Node#%d running periodic HELLO task\n", now, d->nodeID);

    }

  

    // Determine if any active routes exist

    for (k=0; k<NBR_AODV; k++) {

      if (d->routing_table[k].valid) {

	activeroute = 1;

      }

    }

  

    period = ttGetPeriod();

    if (activeroute && (d->lastRREQ < now - period)) {

      if (VERBOSE) {

	mexPrintf("Broadcasting HELLO msg\n");

      }

      

      hellomsg = new HelloMsg;

      hellomsg->hopCnt = 0;

      hellomsg->dest = d->nodeID;

      hellomsg->destSeqNbr = d->seqNbrs[d->nodeID];

      hellomsg->src = 0;

      hellomsg->lifetime = DELETE_PERIOD; 

  

      nwkmsg = new GenericNwkMsg;

      nwkmsg->type = HELLO;

      nwkmsg->intermed = d->nodeID;

      nwkmsg->msg = hellomsg;

      ttSendMsg(0, nwkmsg, 24);

    }

    

    // Determine local connectivity

    for (k=0; k<NBR_AODV; k++) {

      if (d->nbors[k]) {

	// Node k is a neighbor

	if (now - d->lastHello[k] > DELETE_PERIOD) {

	  mexPrintf("Node#%d lost connection to Node#%d\n", d->nodeID, k+1);

                

	  d->nbors[k] = 0; // remove from neighbor list

                

	  // Send RERRs

	  for (m=0; m<NBR_AODV; m++) {

	    // Find routes that use node k as next hop

	    if (d->routing_table[m].valid) {

	      if (d->routing_table[m].nextHop == k+1) {

		// Should send RERR to all nodes in precursor list

		for (n=0; n<NBR_AODV; n++) {

		  if (d->routing_table[m].prec[n]) {

		    // Node n uses this node as next hop towards node k

		    rerrmsg = new RERRMsg;

	                                

		    rerrmsg->dest = m + 1;

		    rerrmsg->destSeqNbr = d->routing_table[m].destSeqNbr;

		    rerrmsg->receiver = n + 1;

	                                

		    d->RERRlist->appendNode(new DataNode(rerrmsg, ""));

		  }

		}

	  

		// Invalidate route

		if (VERBOSE) {

		  mexPrintf("Node#%d invalidating route to Node#%d through unreachable Node#%d\n", d->nodeID, m+1, k+1);

		}

	    

		d->routing_table[m].valid = 0;

	      }

	    }

	  }

	} 

      }

    }

  

    return 0.0001;;

  case 2:

    ttSleep(0.001);

    return 0.0001;    

  case 3:

    // Send all RERRs

    dn = (DataNode*) d->RERRlist->getFirst(); 

    if (dn != NULL) {

      rerrmsg = (RERRMsg*) dn->data;



      nwkmsg = new GenericNwkMsg;

      nwkmsg->type = RERR;

      nwkmsg->intermed = d->nodeID;

      nwkmsg->msg = rerrmsg;

            

      ttSendMsg(rerrmsg->receiver, nwkmsg, 12); 

    

      tmp = dn;

      dn = (DataNode*) dn->getNext();

      tmp->remove();

      delete tmp;

        

      ttSleep(0.001);

      return 0.0001;    

    } else {

      return FINISHED; 

    } 

  case 4:    

    ttSetNextSegment(3);

    return 0.0001; 

  }

  

  return FINISHED; // to supress compilation warnings

}

File: examples/AODV/hellocode.m
function [exectime, data] = hellocode(seg, data)

global routing_table
global seqNbrs
global AODVparams
global testvar
global verbose

switch seg,

 case 1,
  noww = ttCurrentTime;
  
  if verbose
    disp(['Time: ' num2str(noww) ' Node#' num2str(data.nodeID) ' running periodic HELLO task']);
  end
  
  table = routing_table{data.nodeID};
  
  l = length(table);
  
  % Determine active routes
  activeroute = 0;
  hellonodes = [];
  for k = 1:l  
    if (strcmp(table{k}.state, 'valid'))
      activeroute = 1;
      hellonodes = [hellonodes table{k}.nextHop];
    end
  end
  
  period = ttGetPeriod;
  if (activeroute & (data.lastRREQ < noww - period))
    if verbose
      disp('Broadcasting HELLO msg');
    end
      
    RREP.type = 4;
    RREP.hopCnt = 0;
    RREP.dest = data.nodeID;
    RREP.destSeqNbr = seqNbrs(data.nodeID, data.nodeID);
    RREP.src = 0;
    RREP.lifetime = AODVparams.DELETE_PERIOD; % ALLOWED_HELLO_LOSS * HELLO_INTERVAL;
    RREP.intermed = data.nodeID;
  
    ttSendMsg(0, RREP, 24);
    
  end
    
  % Determine local connectivity
  keepInd = [];
  for k = 1:length(data.nbors)
    if (noww - data.nbors{k}.lastHello > AODVparams.DELETE_PERIOD)
      disp(['Node#' num2str(data.nodeID) ' lost connection to Node#' num2str(data.nbors{k}.nborID)]);
      
      % Send RERRs
      for m = 1:length(table)
	if (strcmp(table{m}.state, 'valid') & table{m}.nextHop == data.nbors{k}.nborID)
	  % Should send RERR to all nodes in precursor list
	  for n = 1:length(table{m}.nbor)
	    RERRentry.dest = table{m}.dest;
	    RERRentry.destSeqNbr = table{m}.destSeqNbr;
	    RERRentry.nbor = table{m}.nbor(n);
	    
	    l = length(data.RERRlist);
	    data.RERRlist{l+1} = RERRentry;
	  end
	  
	  % Invalidate route
	  if verbose
	    disp(['Node#' num2str(data.nodeID) ' invalidating route to Node#' num2str(table{m}.dest) ' through unreachable Node#' num2str(data.nbors{k}.nborID)]);
	  end
	    
	  table{m}.state = 'invalid';
	end
      end
      
      % Update table
      routing_table{data.nodeID} = table;
      
    else
      % keep this entry
      keepInd = [keepInd k];
    end
  end
  if (~isempty(keepInd))
    data.nbors = data.nbors(keepInd);
  else
    data.nbors = {};
  end
  
  exectime = 0.0001;
  
 case 2,
  % Send all RERRs
  if (~isempty(data.RERRlist))
    data.cnt = data.cnt + 1;
    
    if (data.cnt > length(data.RERRlist))
      
      data.RERRlist = {}; 
      data.cnt = 0;
      exectime = -1;
      
    else
      
      RERRentry = data.RERRlist{data.cnt};
      RERR.type = 3;
      RERR.dest = RERRentry.dest;
      RERR.destSeqNbr = RERRentry.destSeqNbr;
      RERR.intermed = data.nodeID;
      ttSendMsg(RERRentry.nbor, RERR, 12); 
    
      ttSleep(0.001);
      exectime = 0;
    
    end
    
  else
    exectime = -1;
  end 
  
 case 3,  
  ttSetNextSegment(2);
  exectime = 0;
  
end